### PR TITLE
Apple Silicon support

### PR DESCRIPTION
**Pull request by brainstorm**
This is a followup on PR https://github.com/NationalSecurityAgency/ghidra/pull/2579. Other closely related issues are: https://github.com/NationalSecurityAgency/ghidra/issues/2910 , https://github.com/NationalSecurityAgency/ghidra/pull/2644

The current error as pointed out in https://github.com/NationalSecurityAgency/ghidra/pull/2579#issuecomment-842785570 is the following:

```
% gradle buildGhidra

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':createInstallationZip'.
> Task with path 'assembleDistribution_aarch64' not found in root project 'ghidra'.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/7.0/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 5s
```

This PR needs a bit more work separating `arch` from `OS` detection as pointed out by @ryanmkurtz.

### CODE REVIEW

**Comment by ryanmkurtz** [comment]
The initial roadblock is that we don't want to use aarch64 for a platform name...platforms are operating systems.  It's an issue because we might want to support aarch64 on linux too, for example.  This is what the team needs to decide how to handle properly moving forward.

**Comment by brainstorm** [comment]
@ryanmkurtz 100% agree to that statement. I think that putting effort on CI/CD is absolutely crucial to not only get this right once, but keeping it right on every commit... I see there's a `gitlab-ci.yml`, but more work is needed in CI first.

**Comment by brainstorm** [comment]
I just got a working build with this PR... not perfect but works on my (M1) computer ;)

**Comment by brainstorm** [comment]
After 3 days of using it flawlessly, doing the same things I did on my Intel Mac, I'd merge this, @ryanmkurtz.

It gives Ghidra release 10 a nice entry in the changelog: (experimental?) Apple Silicon Support (AARCH64) ;)

File: GPL/gpl.gradle [hunk 1]
```diff
@@ -45,6 +46,11 @@ ext.getCurrentPlatformName = {
 			return 'osx64'
 		}
 	}
+	else if (osName.startsWith("Mac OS X")) {
+		if (isAARCH_64) {
```
Comment (nickg-ca) [inline GPL/gpl.gradle:50]: Shouldn't this be an "else if (isAARCH_64) within the previous "else if (osName.startsWith("Mac OS X"))"?

Comment (nickg-ca) [inline GPL/gpl.gradle:50]: This (line 50-52) should be changed to an "else if" after the "if (isX86_64)" conditional from line 45-47 I think?

Comment (brainstorm) [reply GPL/gpl.gradle]: Sure thing, looks neater now, thanks!

**Comment by Manouchehri** [comment]
Based on https://github.com/NationalSecurityAgency/ghidra/issues/2653#issuecomment-762994376, it sounds like Ghidra won't ship native M1 support. =/

**Comment by brainstorm** [comment]
You are correct, even if this PR makes the build succeed, there's a few things that do not work well within Ghidra, quite a bit more of integration work is needed, i.e:

```
% gradle buildDependentsDecompile
> Task :Decompiler:compileDecompileLinux64ExecutableDecompileCpp FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':Decompiler:compileDecompileLinux64ExecutableDecompileCpp'.
> No tool chain is available to build for platform 'linux64':
    - Tool chain 'visualCpp' (Visual Studio):
        - Visual Studio is not available on this operating system.
    - Tool chain 'gcc' (GNU GCC):
        - Don't know how to build for platform 'linux64'.
    - Tool chain 'clang' (Clang):
        - Don't know how to build for platform 'linux64'.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/7.0/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 1s
1 actionable task: 1 executed
```

:/

**Comment by ryanmkurtz** [comment]
There's a whole Java-side of things that need to be updated (Platform.java) so Ghidra can find the platform-specific executables.  We have been busy getting the final 10.0 release together, but dealing with arm-based platforms is something on our radar.  As I mentioned earlier though, we need to have some internal discussions to figure out how to best approach it.  I am going to go ahead and close out this PR since it's not something we plan on accepting.  We appreciate your efforts though.  I'm going to mark it as "duplicate effort" since this is something that will come from our team.

**Comment by ryanmkurtz** [comment]
@brainstorm I did some work on this today, and got us building on Linux arm64.  I don't have access to a macOS M1 though...would you might testing out my branch to see how it works?  

https://github.com/ryanmkurtz/ghidra/tree/platforms

I still have a lot of cleanup to do, but was wondering if M1 support is close or not.

Note that I had to add the path to gcc for Linux arm64 for some reason.  Not sure if I'll need something like that for macOS arm64 too:

https://github.com/ryanmkurtz/ghidra/blob/49e763039c576035a5360c47b1c0a9564455a72d/GPL/nativeBuildProperties.gradle#L26-L35

**Comment by nickg-ca** [comment]
@ryanmkurtz I tried it out, and I had to add a similar path to clang on my M1, but after adding this it worked. I haven't tried it on x86 or tried cross compiling.

```
                clang(Clang) {
                        path "/usr/bin/"
                        target("macos_arm_64") {
                                cCompiler.executable = "clang"
                        }   
                }   

```

**Comment by ryanmkurtz** [comment]
@nickg-ca Thanks for testing!  Did the GUI find the decompiler executable?

**Comment by nickg-ca** [comment]
> @nickg-ca Thanks for testing! Did the GUI find the decompiler executable?

Yes, it did.

**Comment by ryanmkurtz** [comment]
Hmmm, the copy in `macos_x86_64` is no good.  I'll have to try to figure out why that is happening.

**Comment by brainstorm** [comment]
> ```
>                 clang(Clang) {
>                         path "/usr/bin/"
>                         target("macos_arm_64") {
>                                 cCompiler.executable = "clang"
>                         }   
>                 }   
> ```

@nickg-ca can you share the diff?

**Comment by ryanmkurtz** [comment]
@nickg-ca @brainstorm could you pull and try again?  I've made some tweaks.

@brainstorm the latest changes are in https://github.com/ryanmkurtz/ghidra/tree/platforms

**Comment by nickg-ca** [comment]
@ryanmkurtz The build works on my M1 with those changes. It looks like it only generated the macos_arm_64 decompile/sleigh as well. One thing I should have mentioned before is that the "ghidraRun" and "support/launch.sh" scripts didn't end up being present in the build.

**Comment by ryanmkurtz** [comment]
Excellent, thanks for testing again.  I see the problem with those missing scripts.  I'm going to clean everything up and I will make sure that is fixed too.

**Comment by ryanmkurtz** [comment]
These changes can now be tracked in #3197.

**Comment by Ruach** [comment]
@nickg-ca Sorry to ask it but where should I those clang things to compile the ghidra?  I got below error 
Task with path 'assembleDistribution_aarch64' not found in root project 'ghidra'

**Comment by Ruach** [comment]
@ryanmkurtz @nickg-ca  I followed the described instructions, but faced below errors. Is there any comments to resolve this ?

**Comment by ryanmkurtz** [comment]
@Ruach, this PR has been closed.  The PR and branch to be testing is in #3197.

**Pull request closed without merging**